File: eArchiveSystem/Domain/Models/Document.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace eArchiveSystem.Domain.Models
{
    [BsonIgnoreExtraElements]
    public class Document
        { 
            // MongoDB ObjectId
            [BsonId]
            [BsonRepresentation(BsonType.ObjectId)]
            public string Id { get; set; }

            // Document title
            public string Title { get; set; }

            // OCR extracted text (optional)
            public string? Content { get; set; }

            // OCR raw text returned by the OCR provider
            public string? RawOcrText { get; set; }

            // OCR normalized text before downstream analysis/search cleaning
            public string? NormalizedOcrText { get; set; }

            // OCR provider metadata
            public string? OcrProvider { get; set; }
            public string? OcrLanguage { get; set; }
            public int? OcrPages { get; set; }
            public DateTime? OcrUpdatedAt { get; set; }

            // File path on the server
            public string FilePath { get; set; }

            // Original file name
            public string FileName { get; set; }

            // File MIME type (e.g. application/pdf)
            public string ContentType { get; set; }

            // File size in bytes
            public long Size { get; set; }

            // File hash (duplicate detection)
            public string FileHash { get; set; }

            // Creation & last update timestamps
            public DateTime CreatedAt { get; set; }
            public DateTime UpdatedAt { get; set; }

            // Owner user ID
            public string UserId { get; set; }

            // Institution that owns this document
            public string? InstitutionId { get; set; }

            // Department identifier used for authorization/search
            public string? DepartmentId { get; set; }

            // Department the document belongs to
            public string? Department { get; set; }

            // Document priority
            [BsonElement("priority")]
            public DocumentPriority Priority { get; set; } = DocumentPriority.Normal;

            // Sensitivity label used by secure download workflows
            [BsonElement("isSensitive")]
            public bool IsSensitive { get; set; } = false;

            // Document workflow status
            [BsonElement("status")]
            public DocumentStatus Status { get; set; } = DocumentStatus.Draft;

            // Workflow timestamps
            [BsonElement("submittedAt")]
            public DateTime? SubmittedAt { get; set; }

            [BsonElement("reviewStartedAt")]
            public DateTime? ReviewStartedAt { get; set; }

            [BsonElement("reviewedAt")]
            public DateTime? ReviewedAt { get; set; }

            [BsonElement("publishedAt")]
            public DateTime? PublishedAt { get; set; }

            [BsonElement("archivedAt")]
            public DateTime? ArchivedAt { get; set; }

            // Workflow actors
            [BsonElement("reviewedByUserId")]
            public string? ReviewedByUserId { get; set; }

            [BsonElement("publishedByUserId")]
            public string? PublishedByUserId { get; set; }

            [BsonElement("archivedByUserId")]
            public string? ArchivedByUserId { get; set; }

            // Rejection reason
            [BsonElement("rejectionReason")]
            public string? RejectionReason { get; set; }

            // Embedded metadata (synced with Metadata collection)
            [BsonElement("Metadata")]
            public Metadata? Metadata { get; set; }

            // Owner name (for view only – not stored in DB)
            [BsonIgnore]
            public string? OwnerName { get; set; }
        }
    }

File: wathiq-frontend/src/types/document.ts
export type DocumentPriority = "Normal" | "Important" | "Urgent" | 0 | 1 | 2;
export type DocumentStatus =
  | "Draft"
  | "Processing"
  | "Submitted"
  | "UnderReview"
  | "Approved"
  | "Rejected"
  | "Published"
  | "Archived"
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7;

export type Owner = {
  id: string;
  name: string;
  email?: string | null;
};

export type Metadata = {
  id?: string | null;
  description?: string | null;
  category?: string | null;
  tags?: string[] | null;
  department?: string | null;
  departmentId?: string | null;
  documentType?: string | null;
  expirationDate?: string | null;
  issuingEntity?: string | null;
  referenceNumber?: string | null;
  documentDate?: string | null;
  insights?: string[] | null;
  hasSignature?: boolean;
  signatures?: string[] | null;
  headers?: string[] | null;
  footers?: string[] | null;
  stamps?: string[] | null;
  rawExtractionJson?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
};

export type Document = {
  id: string;
  documentId?: number | null;
  title?: string | null;
  content?: string | null;
  fileName?: string | null;
  contentType?: string | null;
  size?: number | null;
  fileHash?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  userId?: string | null;
  institutionId?: string | null;
  departmentId?: string | null;
  department?: string | null;
  priority?: DocumentPriority | null;
  isSensitive?: boolean | null;
  status?: DocumentStatus | null;
  submittedAt?: string | null;
  reviewStartedAt?: string | null;
  reviewedAt?: string | null;
  publishedAt?: string | null;
  archivedAt?: string | null;
  reviewedByUserId?: string | null;
  publishedByUserId?: string | null;
  archivedByUserId?: string | null;
  rejectionReason?: string | null;
  ownerName?: string | null;
  owner?: Owner | null;
  metadata?: Metadata | null;
};

export type DocumentTimelineEntry = {
  id?: string | null;
  action?: string | null;
  description?: string | null;
  createdAt?: string | null;
  userId?: string | null;
  userName?: string | null;
};

export type OcrProcessingResponse = {
  status: "processing";
  message?: string | null;
};

export type DocumentOcrText = {
  documentId: string;
  title?: string | null;
  status?: DocumentStatus | null;
  rawText: string;
  normalizedText: string;
  provider?: string | null;
  language?: string | null;
  pages?: number | null;
  extractedAt?: string | null;
};
